Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs

```diff
@@ -111,76 +111,31 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
-                    var candidateChildType = new CandidateChildType
-                    {
-                        Type = childType,
-                        PropertiesAndFields = childTypePropertiesAndFields
-                    };
-                    candidateChildTypes.Add(candidateChildType);
-                }
-            }
 
-            foreach (var candidateChildType in candidateChildTypes)
-            {
-                try
-                {
-                    var deserializedObject = serializer.Deserialize(jsonObject.CreateReader(), candidateChildType.Type);
-                    if (deserializedObject != null)
-                    {
-                        candidateChildType.DeserializedObject = deserializedObject;
-                    }
-                }
-                catch (JsonException)
-                {
-                }
-            }
+            var candidateChildTypes = GetCandidateChildTypes(objectType, jsonProperties).ToList();
+
+            DeserializeCandidates(serializer, jsonObject, candidateChildTypes);
 
             candidateChildTypes = candidateChildTypes.Where(c => c.DeserializedObject != null).ToList();
+
             if (candidateChildTypes.Count == 0)
             {
                 throw new JsonSerializationException(
                     string.Format(CultureInfo.InvariantCulture, "Unable to deserialize to type {0}, value: {1}", objectType, jsonObject));
             }
 
-            if (candidateChildTypes.Count > 1)
+            CandidateChildType matchedChild = null;
+            if (candidateChildTypes.Count == 1)
             {
-                var strictCandidates =
-                    candidateChildTypes.Where(cct => cct.PropertiesAndFields.All(pf => jsonProperties.Contains(pf)))
-                        .ToList();
-
-                if (strictCandidates.Count == 1)
-                {
-                    candidateChildTypes = strictCandidates;
-                }
-                else
-                {
-                    if (strictCandidates.Count > 1)
-                    {
-                        candidateChildTypes = strictCandidates;
-                    }
-
-                    var matchingTypes =
-                        candidateChildTypes.Select(_ => _.Type.FullName)
-                            .Aggregate((running, current) => running + " | " + current);
-                    throw new JsonSerializationException(
-                        string.Format(CultureInfo.InvariantCulture, "The json string can be deserialized into multiple types: {0}, value: {1}", matchingTypes, jsonObject));
-                }                
+                matchedChild = candidateChildTypes.Single();
+            }
+            else if (candidateChildTypes.Count > 1)
+            {
+                matchedChild = SelectBestChildUsingStrictPropertyMatching(candidateChildTypes, jsonObject, jsonProperties);
             }
 
-            return candidateChildTypes.Single().DeserializedObject;
```

Comment:
No biggie but potentially the preceding block could read something like var matchedChild = candidateChildTypes.SingleOrDefault() ?? SelectBest....(). Think about if you like that or not...